File: qoract/record.py
"""Frozen record types. Closed enums. No inference helpers here."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class Verdict(str, Enum):
    HUMAN_AUTHORED = "HUMAN_AUTHORED"
    AGENT_AUTHORED = "AGENT_AUTHORED"
    MIXED = "MIXED"
    UNVERIFIABLE = "UNVERIFIABLE"


class Rollup(str, Enum):
    """Record-level only. Never a fifth inferred authorship state."""

    COMPLETE = "COMPLETE"
    PARTIAL_SURFACES = "PARTIAL_SURFACES"
    UNVERIFIABLE = "UNVERIFIABLE"


class MediaKind(str, Enum):
    PIXELS = "pixels"
    SPEECH = "speech"


FORBIDDEN_SIGNERS = frozenset(
    {"bridge", "operator", "qoresence", "qortroller", "qoract"}
)


@dataclass(frozen=True)
class OutcomeSurface:
    verdict: Verdict
    agent_actuator: str
    kas_commitment: str | None
    kas_verdict: str | None
    hid_bodied_on_host: bool
    ivc_joined: bool


@dataclass(frozen=True)
class MediaSpan:
    kind: MediaKind
    verdict: Verdict
    commit: bool
    armed: bool
    path: str
    source: str
    ticket_id: str
    clock_ns: int
    frame_seq: int | None
    stem: str | None


@dataclass(frozen=True)
class QorActRecord:
    schema: str
    live: bool
    sealed: bool
    session_id: str | None
    session_display: str
    recap_ref: str | None
    clock_commitment: str | None
    outcome: OutcomeSurface | None
    media: tuple[MediaSpan, ...]
    fusion_proof_refs: tuple[str, ...]
    rollup: Rollup
    hygiene_note: str
    honesty: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema": self.schema,
            "live": self.live,
            "sealed": self.sealed,
            "session_id": self.session_id,
            "session_display": self.session_display,
            "recap_ref": self.recap_ref,
            "clock_commitment": self.clock_commitment,
            "outcome": None
            if self.outcome is None
            else {
                "verdict": self.outcome.verdict.value,
                "agent_actuator": self.outcome.agent_actuator,
                "kas_commitment": self.outcome.kas_commitment,
                "kas_verdict": self.outcome.kas_verdict,
                "hid_bodied_on_host": self.outcome.hid_bodied_on_host,
                "ivc_joined": self.outcome.ivc_joined,
            },
            "media": [
                {
                    "kind": s.kind.value,
                    "verdict": s.verdict.value,
                    "commit": s.commit,
                    "armed": s.armed,
                    "path": s.path,
                    "source": s.source,
                    "ticket_id": s.ticket_id,
                    "clock_ns": s.clock_ns,
                    "frame_seq": s.frame_seq,
                    "stem": s.stem,
                }
                for s in self.media
            ],
            "fusion_proof_refs": list(self.fusion_proof_refs),
            "rollup": self.rollup.value,
            "hygiene_note": self.hygiene_note,
            "honesty": dict(self.honesty),
        }

File: glass/src/lib/qoract/record.ts
/** Frozen record types. Closed enums. No inference helpers here. */

export const SCHEMA = "qoract-record-1";
export const OUTCOME_AGENT = "undeployed";

export const Verdict = {
  HUMAN_AUTHORED: "HUMAN_AUTHORED",
  AGENT_AUTHORED: "AGENT_AUTHORED",
  MIXED: "MIXED",
  UNVERIFIABLE: "UNVERIFIABLE",
} as const;
export type Verdict = (typeof Verdict)[keyof typeof Verdict];

export const Rollup = {
  COMPLETE: "COMPLETE",
  PARTIAL_SURFACES: "PARTIAL_SURFACES",
  UNVERIFIABLE: "UNVERIFIABLE",
} as const;
export type Rollup = (typeof Rollup)[keyof typeof Rollup];

export const MediaKind = {
  PIXELS: "pixels",
  SPEECH: "speech",
} as const;
export type MediaKind = (typeof MediaKind)[keyof typeof MediaKind];

export const FORBIDDEN_SIGNERS = new Set([
  "bridge",
  "operator",
  "qoresence",
  "qortroller",
  "qoract",
]);

export type OutcomeSurface = {
  verdict: Verdict;
  agent_actuator: string;
  kas_commitment: string | null;
  kas_verdict: string | null;
  hid_bodied_on_host: boolean;
  ivc_joined: boolean;
};

export type MediaSpan = {
  kind: MediaKind;
  verdict: Verdict;
  commit: boolean;
  armed: boolean;
  path: string;
  source: string;
  ticket_id: string;
  clock_ns: number;
  frame_seq: number | null;
  stem: string | null;
};

export type QorActRecord = {
  schema: string;
  live: boolean;
  sealed: boolean;
  session_id: string | null;
  session_display: string;
  recap_ref: string | null;
  clock_commitment: string | null;
  outcome: OutcomeSurface | null;
  media: MediaSpan[];
  fusion_proof_refs: string[];
  rollup: Rollup;
  hygiene_note: string;
  honesty: Record<string, unknown>;
};

export function recordToDict(rec: QorActRecord): Record<string, unknown> {
  return {
    schema: rec.schema,
    live: rec.live,
    sealed: rec.sealed,
    session_id: rec.session_id,
    session_display: rec.session_display,
    recap_ref: rec.recap_ref,
    clock_commitment: rec.clock_commitment,
    outcome: rec.outcome
      ? {
          verdict: rec.outcome.verdict,
          agent_actuator: rec.outcome.agent_actuator,
          kas_commitment: rec.outcome.kas_commitment,
          kas_verdict: rec.outcome.kas_verdict,
          hid_bodied_on_host: rec.outcome.hid_bodied_on_host,
          ivc_joined: rec.outcome.ivc_joined,
        }
      : null,
    media: rec.media.map((s) => ({
      kind: s.kind,
      verdict: s.verdict,
      commit: s.commit,
      armed: s.armed,
      path: s.path,
      source: s.source,
      ticket_id: s.ticket_id,
      clock_ns: s.clock_ns,
      frame_seq: s.frame_seq,
      stem: s.stem,
    })),
    fusion_proof_refs: [...rec.fusion_proof_refs],
    rollup: rec.rollup,
    hygiene_note: rec.hygiene_note,
    honesty: { ...rec.honesty },
  };
}

export function isVerdict(v: string): v is Verdict {
  return (Object.values(Verdict) as string[]).includes(v);
}

export function isRollup(v: string): v is Rollup {
  return (Object.values(Rollup) as string[]).includes(v);
}

export function isMediaKind(v: string): v is MediaKind {
  return (Object.values(MediaKind) as string[]).includes(v);
}
